### PR TITLE
Integrate Lambda Default Parameters into Synthesized Delegate Types and Lowering Pass

### PR DESCRIPTION
**Pull request by adamperlin**
Allows default values to be attached to the parameters of synthesized delegate types and threaded through the closure lowering pass in order to support code-gen for lambda default parameters. 

Still missing (will be added in a subsequent PR): 
- Diagnostics for invalid target-type conversions between lambdas with default parameters and delegate types
- Synthesizing delegate types for method groups with default parameters (and associated conversion warnings)

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:8946]: We should replicate this assert for default values as well.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:671]: Nit: prefer `not null` instead of `{}` if you don't need to create a variable. Don't let @jaredpar corrupt you.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedIntrinsicOperatorSymbol.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedIntrinsicOperatorSymbol.cs:476]: Please use the parameter name for the `null` value.

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:11604]: Thankfully, none of this manual module reading is necessary. `CompileAndVerify` has a `symbolVerifier` callback that will be passed the `ModuleSymbol` instance after being compiled and re-read. However, I don't think any of this is necessary: verifying the type IL seems fine to me.

Another option, instead of verifying the type IL, would be to use `symbolVerifier` to grab these symbols from the module, and verify the properties about them that we want (the invoke method has the default parameter in the right place, for example).

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:11519]: Dead code?

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:11866]: It would likely better communicate what you were going for by calling this `LambdaDefaultParameterNameMismatch`.

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:11923]: PROTOTYPE instead of TODO

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:11944]: PROTOTYPE instead of TODO

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:12163]: It's worth having a test or two for the following scenarios:

1. Out of order parameter invocation where all the parameters are specified.
2. Out of order parameter invocation where one of the optional parameters is missing.
3. A scenario with a `decimal` default parameter, as these have more complicated constant value handling.

I dont expect any of these to cause issues, but good to test regardless.

**Review by 333fred — COMMENTED** [review]
Done review pass (commit 7)

File: src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/AnonymousTypeField.cs
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/AnonymousTypeField.cs:32]: Any IDE concerns regarding https://github.com/dotnet/roslyn/blob/c18a66e5912ba8ad6183eed6d2496ae8221a52d0/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/SymbolKey/SymbolKey.AnonymousTypeSymbolKey.cs#L65-L67 ? `CreateAnonymousTypeSymbol` doesn't pass `defaultValue` when creating `AnonymousTypeField`

cc @CyrusNajmabadi

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs:113]: Are these reachable now?

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs
Comment (jaredpar) [reply src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:671]: Actually I'd advocate for `is null` / `is not null`. I don't like `x is { }` because it has the flaw of being silently promoted to `true` when `x` is a `struct`. At this point I think `x is null` is equivalent to `x is object` for all the weird cases I cared about. The `x is {}` is the problem case: it's shorter but less precise :smile:

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs
Comment (jaredpar) [reply src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:11519]: Technically zombie code. It's dead yet still in the repo so sort-of alive therefore zombie

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs
Comment (adamperlin) [reply src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:671]: Jared filled me in on the backstory of `x is { }` and I understand the pitfalls now so I will be changing this to `x is not null`

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs
Comment (adamperlin) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs:113]: Yes, see the local function for binding default parameters in the invocation binder here: https://github.com/dotnet/roslyn/blob/622fc4469a09f7511bebefc2462955b5d889de87/src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs#L1359. It accesses those properties, and we may now be passing in `SynthesizedParameterSymbol`  as the parameter

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs:113]: @adamperlin I see, but it seems there is no test coverage for caller info, and returning false doesn't feel correct to me (unless I'm missing something). We need some tests similar to this case anyway:

```csharp
Action<string> x = ([CallerMemberName] x = "") => Console.WriteLine(x);
x();
```

Whether this is something to address in this PR or in a follow-up is the compiler's team decision. I think a PROTOTYPE comment might be sufficient for now.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs:301]: If this assertion failed, you will need to add `|| defaultValue is not null` (or `|| defaultValue is object`)

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs:16]: Unused?

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs
Comment (adamperlin) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs]: Yes this seems reasonable to me, in which case it might be worth renaming `SynthesizedParameterSymbol` to `SynthesizedSimpleParameterSymbol`. I've made that change

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs
Comment (adamperlin) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs]: Yes I initially switched over to `SourceComplexParameterSymbol` and forgot to update those assertions and a few other things, so I reverted. The assertion issue should be fixed now though

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:6]: Unused?

**Review by 333fred — COMMENTED** [review]
Done review pass. Just a couple of small things.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs
Comment (adamperlin) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs]: On second thought, perhaps renaming wouldn't be the clearest, due to the presence of the `Create` factory method.

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs]: @adamperlin The static method is probably better on `SynthesizedParameterSymbolBase`?

@333fred What are your thoughts here?

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs:24]: Let's move this down to `SynthesizedComplexParameterSymbol`, rather than putting it in the base, and override where appropriate (including for the caller info properties).

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs:113]: Let's add these tests now.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs]: What method are you referring to @Youssef1313?

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs]: `Create` factory method

https://github.com/dotnet/roslyn/blob/0f757da6204418a21dbf129b90831f5200a2e388/src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs#L213-L239

Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs]: We didn't create that method here, so let's not move it as part of this pr.

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:11499]: I was expecting this method and all references to be removed.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs
Comment (adamperlin) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs]: @Youssef1313 yes I think it makes more sense in `SynthesizedParameterSymbolBase`, but I won't move it in this pr.

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs
Comment (adamperlin) [reply src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs]: Missed those - sorry

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs
Comment (adamperlin) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs]: Added test coverage for the  `Caller...` attributes. The attributes are currently ignored

Comment (adamperlin) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs]: I have added a PROTOTYPE comment to address this issue in LDM

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs
Comment (Youssef1313) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:12220]: `CallerArgumentExpression` as well?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:8943]: > = default [](http://example.com/codeflow?start=65&length=10)

There are only two callers of `GetMethodGroupOrLambdaDelegateType()`, and both will likely support default values, so we let's remove the default value for this parameter.

**Comment by cston** [comment]
>                     parameterDefaultValueBuilder.Add(constVal);

Consider moving this after the `if (constVal != null) { ... }` check, essentially moving a version of the `if`  below here, to avoid needing the comment.
```csharp
if (constVal != null && parameterDefaultValueBuilder == null)
{
    parameterDefaultValueBuilder = ... // allocate and fill
}
if (parameterDefaultValueBuilder != null)
{
    parameterDefaultValueBuilder.Add(constVal);
}
```

---
Refers to: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:664 in 8254fca. [](commit_id = 8254fcad38c78e9ccb63c99ea20cc47cdaceeb89, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs:77]: >internal override ConstantValue? ExplicitDefaultConstantValue => null; [](http://example.com/codeflow?start=8&length=70)

Consider reverting this if the property hasn't changed.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs:316]: >_baseParameterForAttributes.IsMetadataOptional == true [](http://example.com/codeflow?start=95&length=54)

Is it possible that `_baseParameterForAttributes.ExplicitDefaultConstantValue == null` and `_defaultValue != null`?

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs:353]: Consider reverting.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs
Comment (adamperlin) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs:316]: No, it's actually not, since if `_baseParameterForAttributes` is non-null, then `_baseParameterForAttributes.ExplicitDefaultConstantValue == _defaultValue`. However, we could have a situation where `baseParameterForAttributes == null`, (since it was not passed in) but `_defaultValue != null`

Comment (adamperlin) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs:316]: Should we always attempt to pass in `baseParameterForAttributes` if possible?

File: src/Compilers/CSharp/Portable/Lowering/SynthesizedMethodBaseSymbol.cs
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Lowering/SynthesizedMethodBaseSymbol.cs:57]: nit

```suggestion
```

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs
Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedParameterSymbol.cs]: > I just checked and it shouldn't be possible, because if _baseParameterForAttributes is non-null, then _baseParameterForAttributes.ExplicitDefaultConstantValue == _defaultValue. 

Let's assert that in the constructor. That should be sufficient, thanks.

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:12214]: Nit: formatting of this test appears off.

**Review by 333fred — COMMENTED** [review]
Done review pass. Overall looking good, but test failure needs to be addressed.

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:12244]: >); [](http://example.com/codeflow?start=162&length=2)

Should this include `expectedOutput: "3"`?

**Review by cston — APPROVED** [review]

**Review by 333fred — APPROVED** [review]
LGTM (commit 21), assuming tests are passing.

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs
Comment (adamperlin) [reply src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:12244]: Expected output is actually the default value since the `[CallerArgumentExpression]` attribute isn't supported on lambda parameters. But yes, it does need expected output I just fixed that

File: src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/AnonymousTypeField.cs
Comment (jaredpar) [reply src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/AnonymousTypeField.cs:32]: We should file a prototype comment to track this and sync with IDE team before main merge

File: src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs:6975]: Did we consider having two variations of `VerifyDiagnostics` here: langVersion 11 and langVersion preview?

File: src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs:6993]: Same

File: src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs:6975]: I was planning asking for a dedicated langversion test in the next PR, as it will be focusing on errors :).

File: src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/AnonymousTypeField.cs
Comment (adamperlin) [reply src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/AnonymousTypeField.cs:33]: Just added

File: src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs
Comment (adamperlin) [reply src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs:6975]: Are we ok to wait until a next PR for this or should I do this now?

**Pull request merged**
